casanovo/denovo/model_runner.py

```diff
@@ -394,9 +415,15 @@ def _get_index(
         else:
             index_fname = Path(self.tmp_dir.name) / f"{uuid.uuid4().hex}.hdf5"
 
-        Index = AnnotatedSpectrumIndex if annotated else SpectrumIndex
-        valid_charge = np.arange(1, self.config.max_charge + 1)
-        return Index(index_fname, filenames, valid_charge=valid_charge)
+        try:
```

Comment:
I'm not super happy with this, but unfortunately depth charge raises a `ValueError` exception when the `AnnotatedSpectrumIndex` constructor is called with an unnonotated MGF file is in the peak file list. I figured the best thing to do would to raise a more helpfull exception, but I'm definately open to a better solution.

Reply:
The difference between the `FileNotFoundError`, `ValueError`, and `RuntimeWarning` as checked in the tests also seems a bit weird, because these are essentially the same situations. Can this be made more consistent?

Optionally we can request/implement changes in DepthCharge as well.

Reply:
Another option is that it should be relatively easy to check whether a MGF file is annotated or not, and filter it out if it isn't annotated. A quick implementation I can think of is to do a pass through the file and make sure that the number of annotations is equal to the number of spectra (i.e. the number of lines that begin with "BEGIN IONS" is equal to the number of lines that begin with "SEQ="). This would eliminate the `ValueError` condition, leaving just the `FileNotFoundError` and `RuntimeWarning`. The `FileNotFound` error would still be thrown if no peak files were found,  under the given peak file path(s), and `RuntimeWarning` would be thrown whenever a file is filtered.

I wish there was a more specific class of warning than `RuntimeWarning` to use in this case. I guess we could implement our own warning class, but I'm not sure that would be worth it just for this use case.

Reply:
I suppose another option could be to leave this:

```python
Index = AnnotatedSpectrumIndex if annotated else SpectrumIndex
valid_charge = np.arange(1, self.config.max_charge + 1)
return Index(index_fname, filenames, valid_charge=valid_charge)
```

as is and simply let depth charge throw the `TypeError` exception without catching it. Considering that #207 seems to be a very similar error related to missing `SEQ=` in an annotated MGF file, and was marked as a depth charge v0.4 milestone, imo this isn't a bad way to go - basically treating this specific issue as a depth charge issue.

As for keeping the try/except block and checking to make sure that the `TypeError` is related to trying to construct an `AnnotatedSpectrumIndex` using a unannotated MGF file, I am not quite sure on a good way of approaching this.

Reply:
> As for keeping the try/except block and checking to make sure that the TypeError is related to trying to construct an AnnotatedSpectrumIndex using a unannotated MGF file, I am not quite sure on a good way of approaching this.

You'd need to check the error message, although it's not the most robust approach.

We can discuss desired behavior during the call.

Reply:
I tried running evaluate with a unannotated MGF file using the current dev version of Casanovo, and I got the same error:

```
(casanovo_env) C:\Users\gwen\OneDrive\Documents\casanovo>casanovo evaluate data\sample_unannotated.mgf
WARNING: Dataloader multiprocessing is currently not supported on Windows or MacOS; using only a single thread.
Seed set to 454
INFO: Model weights file C:\Users\gavin\AppData\Local\casanovo\casanovo_v420_v4_2_0.ckpt retrieved from local cache
INFO: Casanovo version 4.1.1.dev15+gcedfaa7.d20240625
INFO: Sequencing and evaluating peptides from:
INFO:   data\sample_unannotated.mgf
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
INFO: Reading 1 files...
data\sample_unannotated.mgf: 128spectra [00:00, 2212.82spectra/s]
Traceback (most recent call last):
  File "D:\anaconda3\envs\casanovo_env\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\anaconda3\envs\casanovo_env\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "D:\anaconda3\envs\casanovo_env\Scripts\casanovo.exe\__main__.py", line 7, in <module>
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\rich_click\rich_command.py", line 367, in __call__
    return super().__call__(*args, **kwargs)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\rich_click\rich_command.py", line 152, in main
    rv = self.invoke(ctx)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\casanovo.py", line 185, in evaluate
    runner.evaluate(annotated_peak_path)
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\denovo\model_runner.py", line 136, in evaluate
    test_index = self._get_index(peak_path, True, "evaluation")
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\denovo\model_runner.py", line 401, in _get_index
    return Index(index_fname, filenames, valid_charge=valid_charge)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 446, in __init__
    super().__init__(
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 104, in __init__
    self.add_file(ms_file)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 231, in add_file
    group.create_dataset(
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\group.py", line 183, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\dataset.py", line 166, in make_new_dset
    dset_id.write(h5s.ALL, h5s.ALL, data)
  File "h5py\\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py\\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py\\h5d.pyx", line 282, in h5py.h5d.DatasetID.write
  File "h5py\\_proxy.pyx", line 147, in h5py._proxy.dset_rw
  File "h5py\\_conv.pyx", line 442, in h5py._conv.str2vlen
  File "h5py\\_conv.pyx", line 96, in h5py._conv.generic_converter
  File "h5py\\_conv.pyx", line 247, in h5py._conv.conv_str2vlen
TypeError: Can't implicitly convert non-string objects to strings
```

The same thing happens with the train command as well:

```
(casanovo_env) C:\Users\gwen\OneDrive\Documents\casanovo>casanovo train data\sample_unannotated.mgf -p data\sample_unannotated.mgf 
WARNING: Dataloader multiprocessing is currently not supported on Windows or MacOS; using only a single thread.
Seed set to 454
INFO: Casanovo version 4.1.1.dev15+gcedfaa7.d20240625
INFO: Training a model from:
INFO:   data\sample_unannotated.mgf
INFO: Using the following validation files:
INFO:   data\sample_unannotated.mgf
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
INFO: Reading 1 files...
data\sample_unannotated.mgf: 128spectra [00:00, 3714.21spectra/s]
Traceback (most recent call last):
  File "D:\anaconda3\envs\casanovo_env\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "D:\anaconda3\envs\casanovo_env\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "D:\anaconda3\envs\casanovo_env\Scripts\casanovo.exe\__main__.py", line 7, in <module>
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\rich_click\rich_command.py", line 367, in __call__
    return super().__call__(*args, **kwargs)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\rich_click\rich_command.py", line 152, in main
    rv = self.invoke(ctx)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\casanovo.py", line 232, in train
    runner.train(train_peak_path, validation_peak_path)
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\denovo\model_runner.py", line 110, in train
    train_index = self._get_index(train_peak_path, True, "training")
  File "C:\Users\gavin\OneDrive\Documents\casanovo\casanovo\denovo\model_runner.py", line 401, in _get_index
    return Index(index_fname, filenames, valid_charge=valid_charge)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 446, in __init__
    super().__init__(
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 104, in __init__
    self.add_file(ms_file)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py", line 231, in add_file
    group.create_dataset(
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\group.py", line 183, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\dataset.py", line 166, in make_new_dset
    dset_id.write(h5s.ALL, h5s.ALL, data)
  File "h5py\\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py\\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py\\h5d.pyx", line 282, in h5py.h5d.DatasetID.write
  File "h5py\\_proxy.pyx", line 147, in h5py._proxy.dset_rw
  File "h5py\\_conv.pyx", line 442, in h5py._conv.str2vlen
  File "h5py\\_conv.pyx", line 96, in h5py._conv.generic_converter
  File "h5py\\_conv.pyx", line 247, in h5py._conv.conv_str2vlen
TypeError: Can't implicitly convert non-string objects to strings
```

Is the error handling for this something we want to have on this PR, or should this be handled separately (namely after #350 is merged in)?

Reply:
I don't think that that will change with the new DepthCharge version, so we can address it here already as well.